File: leetcode/medium/95-unique-bst-2/iter-dp.py
"""
  @author Vortexx2
  Problem 95 - Unique BST 2

  Runtime - 78 ms
  Memory Usage - 17 MB
"""
from typing import *


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def clone(self, root: TreeNode) -> TreeNode:
    if not root:
      return None

    return TreeNode(root.val, self.clone(root.left), self.clone(root.right))

  def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
    res = [None]

    for i in range(1, n + 1):
      temp = []
      for j in range(len(res)):
        oldRoot = res[j]
        root = TreeNode(i)
        root.left = self.clone(oldRoot)
        temp.append(root)

        if oldRoot is not None:
          tempOld = oldRoot

          while tempOld:
            tright = tempOld.right
            newRoot = TreeNode(i)
            tempOld.right = newRoot
            newRoot.left = tright
            temp.append(self.clone(oldRoot))
            tempOld.right = tright
            tempOld = tempOld.right

      res = temp

    return res

File: leetcode/medium/95-unique-bst-2/iter-dp.cpp
/**
 * @file iter-dp.cpp
 * @author Vortexx2
 * @brief Problem 95 - Unique BST 2
 * @date 01-09-2021
 * 
 * Runtime - 16 ms
 * Memory Usage - 13.6 MB
 */
#include <algorithm>
#include <climits>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<TreeNode *> generateTrees(int n) {
    vector<TreeNode *> res = {nullptr};

    for (int i = 1; i <= n; i++) {
      vector<TreeNode *> temp;
      for (int j = 0; j < res.size(); j++) {
        TreeNode *oldRoot = res[j];
        TreeNode *root = new TreeNode(i);
        TreeNode *target = clone(oldRoot);
        root->left = target;
        temp.push_back(root);

        if (oldRoot != nullptr) {
          TreeNode *tempOld = oldRoot;
          while (tempOld) {
            TreeNode *newNode = new TreeNode(i);
            TreeNode *tright = tempOld->right;
            tempOld->right = newNode;
            newNode->left = tright;
            TreeNode *target = clone(oldRoot);
            temp.push_back(target);
            tempOld->right = tright;
            tempOld = tempOld->right;
          }
        }
      }

      res = temp;
    }

    return res;
  }

  TreeNode *clone(TreeNode *root) {
    if (!root) return nullptr;
    TreeNode *newRoot =
        new TreeNode(root->val, clone(root->left), clone(root->right));
    return newRoot;
  }
};